File: JavaScript/95.不同的二叉搜索树-ii.js
/*
 * @lc app=leetcode.cn id=95 lang=javascript
 *
 * [95] 不同的二叉搜索树 II
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {number} n
 * @return {TreeNode[]}
 */
var generateTrees = function (n) {
    if (n === 0) return new Array();
    return generate_trees(1, n);

};

var generate_trees = function (start, end) {
    let trees = new Array();
    if (start > end) {
        trees.push(null);
        return trees;
    }
    for (let i = start; i <= end; i++) {
        let r_trees = generate_trees(i + 1, end);
        let l_trees = generate_trees(start, i - 1);

        for (let r of r_trees) {
            for (let l of l_trees) {
                let t = new TreeNode(i);
                t.left = l;
                t.right = r;
                trees.push(t);
            }
        }

    }

    return trees;
}
    // @lc code=end

File: Java/95.不同的二叉搜索树-ii.java
import java.util.List;

/*
 * @lc app=leetcode.cn id=95 lang=java
 *
 * [95] 不同的二叉搜索树 II
 */

// @lc code=start
/**
 * Definition for a binary tree node. public class TreeNode { int val; TreeNode
 * left; TreeNode right; TreeNode(int x) { val = x; } }
 */
class Solution {
    public List<TreeNode> generateTrees(int n) {
        if(n==0) return new LinkedList<TreeNode>();
        else return generate_trees(1, n);
    }

    private List<TreeNode> generate_trees(int start,int end){
        List<TreeNode> all_trees=new LinkedList<TreeNode>();
        if(start>end) {
            all_trees.add(null);
            return all_trees;
        };
        
        for (int i = start; i < end+1; i++) {
            List<TreeNode> l_trees=generate_trees(start,i-1);
            List<TreeNode> r_trees=generate_trees(i+1,end);
            for (TreeNode  lNode : l_trees) {
                for (TreeNode rNode : r_trees) {
                    TreeNode t=new TreeNode(i);
                    t.left=lNode;
                    t.right=rNode;
                    all_trees.add(t);
                }
                
            }
    }
    return all_trees;
        
    }
}
// @lc code=end
